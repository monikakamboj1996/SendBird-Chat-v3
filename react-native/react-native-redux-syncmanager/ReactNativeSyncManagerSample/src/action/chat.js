const Action = {
  Type: {
    MESSAGE_SELECT: 'message-select',
    MESSAGE_DESELECT_ALL: 'message-deselect-all',
    MESSAGE_ADD: 'message-add',
    MESSAGE_UPDATE: 'message-update',
    MESSAGE_REMOVE: 'message-remove',
    MESSAGE_CLEAR: 'message-clear'
  }
};
Action.selectMessage = message => {
  return {
    type: Action.Type.MESSAGE_SELECT,
    payload: {
      message
    }
  };
};
Action.deselectAllMessages = () => {
  return {
    type: Action.Type.MESSAGE_DESELECT_ALL
  };
};
Action.addMessages = messages => {
  return {
    type: Action.Type.MESSAGE_ADD,
    payload: {
      messages
    }
  };
};
Action.updateMessages = (messages, isTransformFailedToManual) => {
  return {
    type: Action.Type.MESSAGE_UPDATE,
    payload: {
      messages,
      isTransformFailedToManual
    }
  };
};
Action.removeMessages = messages => {
  return {
    type: Action.Type.MESSAGE_REMOVE,
    payload: {
      messages
    }
  };
};
Action.clearMessages = () => {
  return {
    type: Action.Type.MESSAGE_CLEAR
  };
};

export default Action;
