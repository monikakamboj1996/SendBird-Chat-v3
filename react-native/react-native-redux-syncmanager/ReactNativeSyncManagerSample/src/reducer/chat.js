import Action from '../action/chat';
import { findMessageIndex, getMessageIndex } from '../utils';

const _initialState = {
  channel: null,
  messages: [],
  inputMode: 'chat'
};

export const reducer = (state = _initialState, action) => {
  const newState = { ...state };
  switch (action.type) {
    case Action.Type.MESSAGE_SELECT: {
      newState.messages = [...newState.messages];
      for (let i = 0; i < newState.messages.length; i++) {
        newState.messages[i]._selected = newState.messages[i].isIdentical(action.payload.message);
      }
      break;
    }
    case Action.Type.MESSAGE_DESELECT_ALL: {
      newState.messages = [...newState.messages];
      for (let i = 0; i < newState.messages.length; i++) {
        newState.messages[i]._selected = false;
      }
      break;
    }
    case Action.Type.MESSAGE_ADD: {
      newState.messages = [...newState.messages];
      for (let i = 0; i < action.payload.messages.length; i++) {
        const message = action.payload.messages[i];
        const isFailed = message.messageId === 0 && message.requestState === 'failed';
        const index = findMessageIndex(message, newState.messages, isFailed);
        if (index >= 0) {
          newState.messages.splice(index, 0, message);
          if (index > 0) {
            newState.messages[index - 1]._isPreviousMessageSentBySameUser =
              newState.messages[index - 1].sender &&
              newState.messages[index].sender &&
              newState.messages[index - 1].sender.userId === newState.messages[index].sender.userId;
          }
          if (index < newState.messages.length - 1) {
            newState.messages[index]._isPreviousMessageSentBySameUser =
              newState.messages[index].sender &&
              newState.messages[index + 1].sender &&
              newState.messages[index].sender.userId === newState.messages[index + 1].sender.userId;
          }
        }
      }
      break;
    }
    case Action.Type.MESSAGE_UPDATE: {
      newState.messages = [...newState.messages];
      for (let i = 0; i < action.payload.messages.length; i++) {
        const message = action.payload.messages[i];
        const index = getMessageIndex(message, newState.messages);
        if (index >= 0) {
          message._isPreviousMessageSentBySameUser = newState.messages[index]._isPreviousMessageSentBySameUser;
          newState.messages[index] = message;
          if (action.payload.isTransformFailedToManual) {
            newState.messages[index]._isManualResend = true;
          }
        }
      }
      break;
    }
    case Action.Type.MESSAGE_REMOVE: {
      newState.messages = [...newState.messages];
      for (let i = 0; i < action.payload.messages.length; i++) {
        const message = action.payload.messages[i];
        const index = getMessageIndex(message, newState.messages);
        if (index >= 0) {
          newState.messages.splice(index, 1);
          if (index < newState.messages.length - 1) {
            newState.messages[index]._isPreviousMessageSentBySameUser =
              newState.messages[index + 1].sender &&
              newState.messages[index].sender &&
              newState.messages[index + 1].sender.userId === newState.messages[index].sender.userId;
          }
        }
      }
      break;
    }
    case Action.Type.MESSAGE_CLEAR: {
      newState.messages = [];
      break;
    }
  }
  return newState;
};
