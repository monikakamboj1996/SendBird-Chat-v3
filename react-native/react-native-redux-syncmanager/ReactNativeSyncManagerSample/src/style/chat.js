
import { StyleSheet } from 'react-native';

export const style = StyleSheet.create({
	headerLeftContainer: {
		flexDirection: "row",
		paddingLeft: 10
	},
	headerLeftItemContainer: {
		marginLeft: 0,
		marginRight: 0
	},
	headerLeftItem: {
		backgroundColor: 'transparent',
		paddingHorizontal: 10
	},
	headerRightContainer: {
		flexDirection: "row",
		paddingRight: 10
	},
	headerRightItemContainer: {
		marginLeft: 0,
		marginRight: 0
	},
	headerRightItem: {
		backgroundColor: 'transparent',
		paddingHorizontal: 10
	},
	container: {
		flex: 1,
		backgroundColor: '#f3f3f3'
	},
	messageListViewContainer: {
		flex: 24,
		padding: 0,
		margin: 0,
    transform: [{ scaleY: -1 }]
  },
  messageInputContainer: {		
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'row',
		backgroundColor:'#fff',
		paddingHorizontal: 6
	},
	formContainer: {
		flex: 1
	},
  formInputContainer: {
		borderBottomColor: 'transparent'
	},
	formInput: {
		color: '#212529',
		maxHeight: 84
	},
	uploadContainer: {
		flexGrow: 0,
		flexShrink: 0,
		flexBasis: 44
	},
	upload: {
		borderRadius: 6,
		backgroundColor: 'transparent'
	},
	sendContainer: {
		flexGrow: 0,
		flexShrink: 0,
		flexBasis: 44
	},
	send: {
		borderRadius: 6,
		backgroundColor: 'transparent'
	}
});