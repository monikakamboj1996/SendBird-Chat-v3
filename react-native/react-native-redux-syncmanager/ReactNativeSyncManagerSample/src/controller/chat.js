import React from 'react';
import { Alert, BackHandler, Clipboard, Keyboard, View } from 'react-native';
import { Button, Input } from 'react-native-elements';
import { connect } from 'react-redux';

import Icon from 'react-native-vector-icons/Ionicons';
import { FlatList } from 'react-native-gesture-handler';
import Permissions from 'react-native-permissions';
import ImagePicker from 'react-native-image-picker';

import SendBird from 'sendbird';
import SendBirdSyncManager from 'sendbird-syncmanager';

import { style } from '../style/chat';
import Action from '../action/chat';
import MessageView from '../view/message';
import { createChannelTitle, toast } from '../utils';
import { navigator } from '../navigator';

const PAGE_RETURN_DELAY = 600;
let componentContext = null;

const _chatCloseHeader = navigation => {
  return (
    <View style={style.headerLeftContainer}>
      <Button
        containerStyle={style.headerLeftItemContainer}
        buttonStyle={style.headerLeftItem}
        icon={<Icon name="md-arrow-back" color="#333" size={28} />}
        onPress={() => {
          navigation.goBack();
        }}
      />
    </View>
  );
};
const _chatControlHeader = navigation => {
  const channel = navigation.getParam('channel', null);
  return (
    <View style={style.headerRightContainer}>
      <Button
        containerStyle={style.headerRightItemContainer}
        buttonStyle={style.headerRightItem}
        icon={<Icon name="md-contacts" color="#7d62d9" size={28} />}
        onPress={() => {
          const sb = SendBird.getInstance();
          sb.GroupChannel.getChannel(channel.url, (channel, err) => {
            if (!err) {
              navigation.push('Member', { channel });
            } else {
              toast('Failed to get channel information.');
            }
          });
        }}
      />
      <Button
        containerStyle={style.headerRightItemContainer}
        buttonStyle={style.headerRightItem}
        icon={<Icon name="md-exit" color="#7d62d9" size={28} />}
        onPress={() => {
          Alert.alert('Leave', 'Do you want to leave this channel?', [
            { text: 'Cancel' },
            {
              text: 'OK',
              onPress: () => {
                channel.leave((res, err) => {
                  if (!err) {
                    navigation.goBack();
                  } else {
                    toast('Failed to leave channel.');
                  }
                });
              }
            }
          ]);
        }}
      />
    </View>
  );
};
const _menuCloseHeader = navigation => {
  const currentChat = navigation.getParam('currentChat', null);
  return (
    <View style={style.headerLeftContainer}>
      <Button
        containerStyle={style.headerLeftItemContainer}
        buttonStyle={style.headerLeftItem}
        icon={<Icon name="md-close" color="#333" size={28} />}
        onPress={() => {
          if (currentChat) {
            currentChat.closeMessageControl();
          } else {
            navigation.setParams({
              currentChat: null,
              selectedMessages: []
            });
          }
        }}
      />
    </View>
  );
};
const _menuControlHeader = navigation => {
  const currentChat = navigation.getParam('currentChat', null);
  const syncManager = SendBirdSyncManager.getInstance();
  const channel = navigation.getParam('channel', null);
  const selectedMessages = navigation.getParam('selectedMessages', []);
  return (
    <View style={style.headerRightContainer}>
      {selectedMessages.length === 1 &&
        selectedMessages[0].isUserMessage() &&
        selectedMessages[0].sender.userId === syncManager.currentUserId &&
        selectedMessages[0].messageId !== 0 && (
          <Button
            containerStyle={style.headerRightItemContainer}
            buttonStyle={style.headerRightItem}
            icon={<Icon name="md-create" color="#7d62d9" size={28} />}
            onPress={() => {
              if (currentChat) {
                currentChat.closeMessageControl();
              } else {
                navigation.setParams({
                  currentChat: null,
                  selectedMessages: []
                });
              }
              currentChat.switchToEditMode(selectedMessages[0]);
            }}
          />
        )}
      {selectedMessages.length === 1 && selectedMessages[0].isUserMessage() && (
        <Button
          containerStyle={style.headerRightItemContainer}
          buttonStyle={style.headerRightItem}
          icon={<Icon name="md-copy" color="#7d62d9" size={28} />}
          onPress={() => {
            if (selectedMessages.length > 0) {
              const message = selectedMessages[0];
              if (message.isUserMessage()) {
                Clipboard.setString(message.message);
                toast('The message is copied to Clipboard.');
                if (currentChat) {
                  currentChat.closeMessageControl();
                } else {
                  navigation.setParams({
                    currentChat: null,
                    selectedMessages: []
                  });
                }
              }
            }
          }}
        />
      )}
      {!selectedMessages[0].isAdminMessage() && selectedMessages[0].sender.userId === syncManager.currentUserId && (
        <Button
          containerStyle={style.headerRightItemContainer}
          buttonStyle={style.headerRightItem}
          icon={<Icon name="md-trash" color="#7d62d9" size={28} />}
          onPress={() => {
            Alert.alert('Delete', 'Do you want to delete this message?', [
              { text: 'Cancel' },
              {
                text: 'OK',
                onPress: () => {
                  for (let i = 0; i < selectedMessages.length; i++) {
                    if (selectedMessages[i].messageId === 0 && selectedMessages[i].requestState !== 'succeeded') {
                      componentContext.collection.deleteMessage(selectedMessages[i]);
                    } else {
                      channel.deleteMessage(selectedMessages[i], (res, err) => {
                        if (err) {
                          console.error(err);
                        }
                      });
                    }
                  }
                  navigation.setParams({
                    currentChat: null,
                    selectedMessages: []
                  });
                }
              }
            ]);
          }}
        />
      )}
    </View>
  );
};

export class ChatController extends React.Component {
  static navigationOptions = ({ navigation }) => {
    const channel = navigation.getParam('channel', null);
    const selectedMessages = navigation.getParam('selectedMessages', []);
    const isMenuControl = selectedMessages.length > 0;
    return {
      title: isMenuControl ? 'Message Selected' : `${createChannelTitle(channel)} (${channel.memberCount})`,
      headerLeft: isMenuControl ? _menuCloseHeader(navigation) : _chatCloseHeader(navigation),
      headerRight: isMenuControl ? _menuControlHeader(navigation) : _chatControlHeader(navigation)
    };
  };

  static getDerivedStateFromProps(props, state) {
    state.messages = [...props.messages];
    return state;
  }

  constructor(props) {
    super(props);
    this.collection = null;
    this.channel = this.props.navigation.getParam('channel', null);
    this.messageInput = null;
    this.state = {
      typedText: '',
      editedMessage: null,
      inputMode: 'chat',
      messages: []
    };
    componentContext = this;
  }

  _messageEventHandler(messages, action, reason) {
    switch (action) {
      case 'insert': {
        this.props.addMessages(messages);
        break;
      }
      case 'update': {
        if (reason === SendBirdSyncManager.MessageCollection.FailedMessageEventActionReason.UPDATE_RESEND_FAILED) {
          this.props.updateMessages(messages, true);
        } else {
          this.props.updateMessages(messages);
        }
        break;
      }
      case 'remove': {
        this.props.removeMessages(messages);
        break;
      }
      case 'clear': {
        this.props.clearMessages();
        break;
      }
    }
  }

  componentDidMount() {
    const sb = SendBird.getInstance();
    const channelHandler = new sb.ChannelHandler();
    channelHandler.onChannelChanged = channel => {
      this.updateChannel(channel);
    };
    channelHandler.onUserJoined = (channel, user) => {
      this.updateChannel(channel);
    };
    channelHandler.onUserLeft = (channel, user) => {
      if (sb.currentUser.userId === user.userId) {
        this.props.navigation.replace('NotFound', { channelUrl: channel.url });
      } else {
        this.updateChannel(channel);
      }
    };
    sb.addChannelHandler(`chat-${this.channel.url}`, channelHandler);

    this.collection = new SendBirdSyncManager.MessageCollection(this.channel);
    this.collection.limit = 50;

    const collectionHandler = new SendBirdSyncManager.MessageCollection.CollectionHandler();
    collectionHandler.onSucceededMessageEvent = (messages, action, reason) => {
      this._messageEventHandler(messages, action, reason);
    };
    collectionHandler.onFailedMessageEvent = (messages, action, reason) => {
      this._messageEventHandler(messages, action, reason);
    };
    collectionHandler.onPendingMessageEvent = (messages, action, reason) => {
      this._messageEventHandler(messages, action, reason);
    };
    this.collection.setCollectionHandler(collectionHandler);
    this.channel.markAsRead();

    SendBirdSyncManager.loggerLevel = 98765;
    this.collection.fetchSucceededMessages('prev', err => {
      this.collection.fetchSucceededMessages('next', err => {
        this.collection.fetchFailedMessages(err => {});
      });
    });

    navigator.push(this.channel.url);
    BackHandler.addEventListener('hardwareBackPress', this.handleBackPress);
  }

  componentWillUnmount() {
    componentContext = null;
    this.props.clearMessages();
    if (this.collection) {
      this.collection.remove();
      this.collection = null;
    }
    navigator.pop();
    BackHandler.removeEventListener('hardwareBackPress', this.handleBackPress);
  }

  handleBackPress = () => {
    const selectedMessages = this.props.navigation.getParam('selectedMessages', []);
    if (selectedMessages.length > 0) {
      this.closeMessageControl();
      return true;
    } else if (this.state.inputMode === 'edit') {
      this.switchToChatMode();
      return true;
    }
    return false;
  };

  updateChannel(channel) {
    this.channel = channel;
    this.props.navigation.setParams({
      title: `${createChannelTitle(channel)} (${channel.memberCount})`
    });
  }

  loadPrevious() {
    this.collection.fetchSucceededMessages('prev', err => {
      if (err) {
        console.log(err);
      }
    });
  }

  sendMessage() {
    if (this.state.typedText) {
      const pendingMessage = this.channel.sendUserMessage(this.state.typedText, (message, err) => {
        // To preserve order of pending messages being processed before same failed ones.
        setTimeout(() => {
          if (this.collection) {
            this.collection.handleSendMessageResponse(err, message);
          }
        }, 100);
      });
      if (this.collection) {
        this.collection.appendMessage(pendingMessage);
        this.setState({ typedText: '' });
      }
    }
  }

  updateMessage() {
    if (this.state.editedMessage && this.state.typedText) {
      const message = this.state.editedMessage;
      if (message.isUserMessage() && message.message !== this.state.typedText) {
        const messageContent = this.state.typedText;
        this.channel.updateUserMessage(
          message.messageId,
          messageContent,
          message.data,
          message.customType,
          (updatedMessage, err) => {
            if (!err) {
              this.collection.updateMessage(updatedMessage);
              this.switchToChatMode();
              Keyboard.dismiss();
            } else {
              toast('Failed to update the message.');
            }
          }
        );
      } else {
        this.switchToChatMode();
        Keyboard.dismiss();
      }
    } else {
      toast('Cannot update the message with empty content.');
    }
  }

  sendImage(source, options = {}) {
    setTimeout(() => {
      const tempMessage = this.channel.sendFileMessage(source, '', '', options.thumbnailSizes, (message, err) => {
        if (!err) {
          if (message && this.collection) {
            this.collection.appendMessage(message);
          }
        } else {
          toast('Failed to send image. ' + err.message);
          if (tempMessage && this.collection) {
            this.collection.deleteMessage(tempMessage);
          }
        }
      });
      if (this.collection) {
        this.collection.appendMessage(tempMessage);
      }
      this.setState({ typedText: '' });
    }, PAGE_RETURN_DELAY);
  }

  pickImage() {
    Permissions.check('photo').then(response => {
      if (response === 'authorized') {
        ImagePicker.showImagePicker(
          {
            title: 'Select Image',
            mediaType: 'photo',
            noData: true
          },
          response => {
            if (!response.didCancel && !response.error && !response.customButton) {
              const source = { uri: response.uri };
              if (response.name) {
                source['name'] = response.fileName;
              } else {
                paths = response.uri.split('/');
                source['name'] = paths[paths.length - 1];
              }
              if (response.type) {
                source['type'] = response.type;
              } else {
                /** For react-native-image-picker library doesn't return type in iOS,
                 *  it is necessary to force the type to be an image/jpeg (or whatever you're intended to be).
                 */
                if (Platform.OS === 'ios') {
                  source['type'] = 'image/jpeg';
                }
              }
              const thumbnailSizes = [{ maxWidth: 200, maxHeight: 200 }];
              this.sendImage(source, { thumbnailSizes });
            }
          }
        );
      } else if (response === 'undetermined') {
        Permissions.request('photo').then(response => {
          if (response === 'authorized') {
            this.pickImage();
          }
        });
      } else {
        toast('Permission denied. You declined the permission to access your photo.');
      }
    });
  }

  showMessageControl(message) {
    const syncManager = SendBirdSyncManager.getInstance();
    if (message.isUserMessage() || (message.isFileMessage() && message.sender.userId === syncManager.currentUserId)) {
      this.props.selectMessage(message);
      this.props.navigation.setParams({
        currentChat: this,
        selectedMessages: [message]
      });
    }
  }

  resendFailedMessage(message) {
    const pendingMessage = this.channel.resendUserMessage(message, (message, err) => {
      if (this.collection) {
        this.collection.handleSendMessageResponse(err, message);
      }
    });
  }

  closeMessageControl() {
    this.props.deselectAllMessages();
    this.props.navigation.setParams({
      currentChat: null,
      selectedMessages: []
    });
  }

  switchToChatMode() {
    this.setState({
      inputMode: 'chat',
      editedMessage: null,
      typedText: ''
    });
  }

  switchToEditMode(message) {
    this.setState({
      inputMode: 'edit',
      editedMessage: message,
      typedText: message.message
    });
    this.messageInput.focus();
  }

  render() {
    const mode = {
      chat: {
        left: {
          icon: 'md-cloud-upload',
          onPress: this.pickImage.bind(this)
        },
        right: {
          icon: 'md-send',
          onPress: this.sendMessage.bind(this)
        }
      },
      edit: {
        left: {
          icon: 'md-close',
          onPress: this.switchToChatMode.bind(this)
        },
        right: {
          icon: 'md-checkmark',
          onPress: this.updateMessage.bind(this)
        }
      }
    };
    const currentMode = mode[this.state.inputMode];
    return (
      <View style={style.container}>
        <View style={style.messageListViewContainer}>
          <FlatList
            data={this.state.messages}
            keyExtractor={message => (message.messageId > 0 ? message.messageId : message.reqId) + ''}
            onEndReached={this.loadPrevious.bind(this)}
            renderItem={bundle => (
              <MessageView
                channel={this.channel}
                message={bundle.item}
                onLongPress={this.showMessageControl.bind(this)}
                resendFailedMessage={this.resendFailedMessage.bind(this)}
              />
            )}
          />
        </View>
        <View style={style.messageInputContainer}>
          <Button
            containerStyle={style.uploadContainer}
            buttonStyle={style.upload}
            icon={<Icon name={currentMode.left.icon} size={28} color="#333" />}
            onPress={currentMode.left.onPress}
          />
          <Input
            placeholder="Your message"
            ref={input => {
              this.messageInput = input;
            }}
            containerStyle={style.formContainer}
            inputContainerStyle={style.formInputContainer}
            inputStyle={style.formInput}
            multiline={true}
            numberOfLines={3}
            autoCapitalize="none"
            autoCorrect={false}
            selectionColor="#ddd"
            underlineColorAndroid="transparent"
            value={this.state.typedText}
            onChangeText={typedText => this.setState({ typedText })}
          />
          <Button
            containerStyle={style.sendContainer}
            buttonStyle={style.send}
            icon={
              <Icon
                name={currentMode.right.icon}
                size={28}
                color={this.state.typedText.length > 0 ? '#7d62d9' : '#494e57'}
              />
            }
            onPress={currentMode.right.onPress}
          />
        </View>
      </View>
    );
  }
}

export default connect(
  previousState => {
    const { chat } = previousState;
    return { ...chat };
  },
  Action
)(ChatController);
