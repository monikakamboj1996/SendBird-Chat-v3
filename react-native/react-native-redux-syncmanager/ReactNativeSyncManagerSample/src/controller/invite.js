import React from 'react';
import { View } from 'react-native';
import { Button } from 'react-native-elements';
import { connect } from 'react-redux';
import Icon from 'react-native-vector-icons/Ionicons';
import { FlatList } from 'react-native-gesture-handler';

import SendBird from 'sendbird';

import { style } from '../style/invite';
import Action from '../action/invite';
import UserView from '../view/user';
import { toast } from '../utils';

let _selectedUserIds = [];

export class InviteController extends React.Component {
  static navigationOptions = ({ navigation }) => {
    const channel = navigation.getParam('channel', null);
    return {
      title: !channel ? 'Create new channel' : 'Invite',
      headerRight: (
        <View style={{ flexDirection: 'row' }}>
          <Button
            containerStyle={style.headerInviteContainer}
            buttonStyle={style.headerInvite}
            icon={<Icon name="md-add" color="#7d62d9" size={28}></Icon>}
            onPress={() => {
              if (!channel) {
                if (_selectedUserIds.length > 0) {
                  const sb = SendBird.getInstance();
                  const params = new sb.GroupChannelParams();
                  params.addUserIds(_selectedUserIds);
                  sb.GroupChannel.createChannel(params, (channel, err) => {
                    if (!err) {
                      navigation.replace('Chat', { channel });
                    } else {
                      toast('Failed to create channel.');
                    }
                  });
                } else {
                  toast('Choose at least 1 user to invite.');
                }
              } else {
                channel.inviteWithUserIds(_selectedUserIds, (res, err) => {
                  if (!err) {
                    navigation.pop();
                  } else {
                    toast('Failed to invite users.');
                  }
                  isLoading = false;
                });
              }
            }}
          />
        </View>
      )
    };
  };
  constructor(props) {
    super(props);
    const sb = SendBird.getInstance();
    this.query = sb.createUserListQuery();
    this.channel = this.props.navigation.getParam('channel', null);
    this.state = {
      users: []
    };
  }
  static getDerivedStateFromProps(props, state) {
    state.users = props.users;
    return state;
  }
  componentDidMount() {
    this.next();
  }
  componentWillUnmount() {
    _selectedUserIds = [];
  }
  toggle(selected, user) {
    if (selected) {
      if (_selectedUserIds.indexOf(user.userId) < 0) {
        _selectedUserIds.push(user.userId);
      }
    } else {
      const index = _selectedUserIds.indexOf(user.userId);
      if (index >= 0) {
        _selectedUserIds.splice(index, 1);
      }
    }
  }
  next() {
    if (this.query.hasNext) {
      this.query.next((users, err) => {
        if (!err) {
          let filteredUsers = [];
          if (this.channel) {
            for (let i in users) {
              if (this.channel.members.map(m => m.userId).indexOf(users[i].userId) < 0) {
                filteredUsers.push(users[i]);
              }
            }
          } else {
            filteredUsers = users;
          }
          if (filteredUsers.length < this.query.limit && this.query.hasNext) {
            setTimeout(() => this.next(), 200);
          } else {
            this.props.fetchUsers(filteredUsers);
          }
        }
      });
    }
  }
  render() {
    return (
      <View>
        <FlatList
          data={this.state.users}
          keyExtractor={user => user.userId + ''}
          onEndReached={this.next.bind(this)}
          renderItem={bundle => (
            <UserView
              user={bundle.item}
              selectable={true}
              selected={_selectedUserIds.indexOf(bundle.item.userId) >= 0}
              onPress={this.toggle.bind(this)}
            />
          )}
        />
      </View>
    );
  }
}

export default connect(
  previousState => {
    const { invite } = previousState;
    return { ...invite };
  },
  Action
)(InviteController);
