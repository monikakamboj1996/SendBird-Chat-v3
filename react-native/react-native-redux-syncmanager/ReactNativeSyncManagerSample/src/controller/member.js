import React from 'react';
import { View } from 'react-native';
import { Button } from 'react-native-elements';
import { connect } from 'react-redux';
import Icon from 'react-native-vector-icons/Ionicons';
import { FlatList } from 'react-native-gesture-handler';

import SendBird from 'sendbird';

import { style } from '../style/member';
import Action from '../action/member';
import UserView from '../view/user';

export class MemberController extends React.Component {
  static navigationOptions = ({ navigation }) => {
    const channel = navigation.getParam('channel', null);
    return {
      title: 'Members',
      headerRight: (
        <View style={{ flexDirection: 'row' }}>
          <Button
            containerStyle={style.headerInviteContainer}
            buttonStyle={style.headerInvite}
            icon={<Icon name="md-add" color="#7d62d9" size={28}></Icon>}
            onPress={() => {
              navigation.push('Invite', { channel });
            }}
          />
        </View>
      )
    };
  };
  constructor(props) {
    super(props);
    this.channel = this.props.navigation.getParam('channel', null);
    this.state = {
      members: []
    };
  }
  static getDerivedStateFromProps(props, state) {
    state.members = props.members;
    return state;
  }
  componentDidMount() {
    const sb = SendBird.getInstance();
    const channelHandler = new sb.ChannelHandler();
    channelHandler.onUserJoined = (channel, user) => {
      this.updateChannel(channel);
    };
    channelHandler.onUserLeft = (channel, user) => {
      this.updateChannel(channel);
    };
    sb.addChannelHandler('member', channelHandler);
    this.props.initMembers(this.channel.members);
  }
  updateChannel(channel) {
    this.channel = channel;
    this.props.updateMembers(this.channel.members);
  }
  render() {
    return (
      <View>
        <FlatList
          data={this.state.members}
          keyExtractor={member => member.userId + ''}
          renderItem={bundle => <UserView user={bundle.item} selectable={false} />}
        />
      </View>
    );
  }
}

export default connect(
  previousState => {
    const { member } = previousState;
    return { ...member };
  },
  Action
)(MemberController);
