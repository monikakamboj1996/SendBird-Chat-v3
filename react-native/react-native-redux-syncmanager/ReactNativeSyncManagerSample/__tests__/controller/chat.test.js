import 'react-native';
import React from 'react';
import { ChatController } from '../../src/controller/chat';
import renderer from 'react-test-renderer';

const navigation = {
  getParam: jest.fn().mockReturnValue({
    url: 'new Url',
    markAsRead: jest.fn()
  }),
  setParams: jest.fn(),
  navigate: jest.fn(),
  state: {
    params: {
      channelUrl: 'channelUrl',
      isOpenChannel: true
    }
  },
  dispatch: jest.fn(),
  addListener: jest.fn()
};

describe('controller/chat', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<ChatController navigation={navigation} failedMessages={[]} messages={[]} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  // TODO: add test cases
});
