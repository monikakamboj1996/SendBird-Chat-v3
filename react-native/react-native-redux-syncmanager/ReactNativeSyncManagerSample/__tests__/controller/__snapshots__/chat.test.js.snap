// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`controller/chat renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#f3f3f3",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 24,
        "margin": 0,
        "padding": 0,
        "transform": Array [
          Object {
            "scaleY": -1,
          },
        ],
      }
    }
  >
    <RCTScrollView
      collapsable={false}
      data={Array []}
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onContentSizeChange={[Function]}
      onEndReached={[Function]}
      onEndReachedThreshold={2}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      renderScrollComponent={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View />
    </RCTScrollView>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingHorizontal": 6,
      }
    }
  >
    <View
      style={
        Object {
          "borderRadius": 6,
          "flexBasis": 44,
          "flexGrow": 0,
          "flexShrink": 0,
        }
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderColor": "#2089dc",
              "borderRadius": 6,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#333",
                  "fontSize": 28,
                },
                undefined,
                Object {
                  "fontFamily": "Ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomColor": "transparent",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          autoCapitalize="none"
          autoCorrect={false}
          multiline={true}
          numberOfLines={3}
          onChangeText={[Function]}
          placeholder="Your message"
          rejectResponderTermination={true}
          selectionColor="#ddd"
          style={
            Object {
              "alignSelf": "center",
              "color": "#212529",
              "flex": 1,
              "fontSize": 18,
              "maxHeight": 84,
              "minHeight": 40,
            }
          }
          testID="RNE__Input__text-input"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
    <View
      style={
        Object {
          "borderRadius": 6,
          "flexBasis": 44,
          "flexGrow": 0,
          "flexShrink": 0,
        }
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderColor": "#2089dc",
              "borderRadius": 6,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#494e57",
                  "fontSize": 28,
                },
                undefined,
                Object {
                  "fontFamily": "Ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
